apm-lambda-extension/apmproxy/apmserver.go

```diff
@@ -210,11 +222,46 @@ func (c *Client) ComputeGracePeriod() time.Duration {
 
 // EnqueueAPMData adds a AgentData struct to the agent data channel, effectively queueing for a send
 // to the APM server.
-func (c *Client) EnqueueAPMData(agentData AgentData) {
+func (c *Client) EnqueueAPMData(agentData AgentData) bool {
 	select {
 	case c.DataChannel <- agentData:
 		c.logger.Debug("Adding agent data to buffer to be sent to apm server")
+		return true
 	default:
 		c.logger.Warn("Channel full: dropping a subset of agent data")
+		return false
 	}
 }
+
+// ShouldFlush returns true if the client should flush APM data after processing the event.
+func (c *Client) ShouldFlush() bool {
+	return c.sendStrategy == SyncFlush || c.hasPendingFlush()
```

Comment:
Looking at this again, I don't know that this is desirable.

I think the `?flushed=true` wording is a bit confusing: it doesn't mean that the extension should flush immediately, it just means that the agent (client) has flushed, which in turn means that the Lambda invocation has completed.

I think we should revert to only synchronously flushing when sendStrategy == SyncFlush.

Reply:
I might be misinterpreting the specification, but I think the goal of `flushing=true` was to reduce latency so that the lambda knows that it can flush its data.

The way I interpreted it was:
- `SyncFlush`: flush on every intake request. We forwards every request from the agent as soon as we receive it.
- `flushed=true`: sent with the final intake request. We buffered the previous requests and this is a signal that we can flush the data.

> I think we should revert to only synchronously flushing when sendStrategy == SyncFlush.

Question: Wouldn't that mean that buffered data have a chance of being flushed only on shutdown or while processing an event ? Is that intended ?
I think that would lead to a potential delay since if the buffer is not emptied we would have to wait for shutdown which could take a while.

Reply:
This is why I think the naming is confusing :)

There's two distinct "flush" events:
 1. agent flushes data to the extension
 2. extension flushes data to the server

The `?flushed=true` request indicates to the extension that the first event has happened. This is a _prerequisite_ for the extension flushing data to the server, but it does not mean the server must flush immediately. This behaviour is intended to be controlled by the [send strategy](https://github.com/elastic/apm-aws-lambda/blob/main/docs/monitoring-aws-lambda.asciidoc#elastic_apm_send_strategy).

> Question: Wouldn't that mean that buffered data have a chance of being flushed only on shutdown or while processing an event ? Is that intended ?
> I think that would lead to a potential delay since if the buffer is not emptied we would have to wait for shutdown which could take a while.

Yes, that is intended. There's a trade-off, as explained in the send strategy docs linked above:
 - In `syncflush` mode, data gets flushed immediately after an invocation. This means the extension cannot service another Lambda invocation until the events are flushed, reducing Lambda invocation throughput.
 - In `background` mode, data gets flushed in the background while subsequent Lambda invocations are being processed, or on shutdown if there are no subsequent invocations. This means that data might be significantly delayed in case there are only sporadic Lambda invocations, but Lambda invocation throughput will not be significantly reduced.

Reply:
Ah I see, thank you for the explanation! 🙇 

I've updated the code to the correct behaviour